README.md
# MeemVite token

A MeemVite token connects you directly with the MeemDAO team on Discord as a full "Meember".

Join us in building the future of digital content where creators set the rules: [https://discord.gg/5NP8PYN8](https://discord.gg/5NP8PYN8)

## Contract addresses

### Polygon (MATIC) Mainnet

MeemVite Proxy: [0xF503e9A36402F9184211b7DCAf097b5a59923bAF](https://polygonscan.com/address/0xF503e9A36402F9184211b7DCAf097b5a59923bAF)

MeemVite Implementation: [0x824b166E7F6289518c8a694CF6385941a85d58DE](https://polygonscan.com/address/0x824b166E7F6289518c8a694CF6385941a85d58DE)

MeemViteURI: [0xe00400D4b6dd34ACCa34b591F395379B65248503](https://polygonscan.com/address/0xe00400D4b6dd34ACCa34b591F395379B65248503)

Opensea: [https://opensea.io/collection/meemvite](https://opensea.io/collection/meemvite)

### Rinkeby Testnet

MeemVite Proxy: [0x275EeF8f111D2B79cB9C196133d1E1aD0a8710F2](https://rinkeby.etherscan.io/address/0x275EeF8f111D2B79cB9C196133d1E1aD0a8710F2)

MeemVite Implementation: [0x0bfae9bb23c5c7633f37af2928b2ffd5594f4809](https://rinkeby.etherscan.io/address/0x0bfae9bb23c5c7633f37af2928b2ffd5594f4809)

MeemViteURI: [0x1d2fA012f7557025Ac9FE1ADB0D738C4Bf57c1E2](https://rinkeby.etherscan.io/address/0x1d2fA012f7557025Ac9FE1ADB0D738C4Bf57c1E2)

Opensea: [https://testnets.opensea.io/collection/meemvite-testnet](https://testnets.opensea.io/collection/meemvite-testnet)

## Metadata

All metadata is 100% on-chain!

* Metadata standards for opensea: https://docs.opensea.io/docs/metadata-standards

* Contract (opensea collection) standards for opensea: https://docs.opensea.io/docs/contract-level-metadata

## Development

By default all commands will use the local network. For other networks use the ```--network <network_name>``` flag. See the hardhat.config.ts file for network names.

### Set up your .env

Copy the `.env.example` file to .env

### Install dependencies

```yarn```

### Compile contracts

```yarn build```

### Watch and compile files automatically

```yarn watch```

### Run tests

```yarn test```

### Run local blockchain

This will start up a local node using hardhat

```yarn network```

**NEVER SEND ETH TO THESE ADDRESSES EXCEPT ON YOUR LOCAL NETWORK**

## Smart Contract Interaction

> **Change the network**
>
> For (deploy, upgrade, console, etc.) commands, you can change the network with `--network <network name>`
>
> The local network is used by default.

### Deploy contract

**You should only do this the first time. After that you should use upgrade to keep the same address**

#### Deploy MeemVite contract

```yarn deploy```

#### Deploy MeemVite URI contract

```yarn deployURI```

### Upgrade the contract

```yarn upgradeContract --contractaddress <address>```

## Console Interaction

This will open a hardhat console where you can interact directly with the smart contract

```yarn console```

### Get a meem instance for use in hardhat console

```
const meemVite = await (await ethers.getContractFactory('MeemVite')).attach('<Contract_address>')
```

### Initialize meem instance with default admin address

When running locally, initialize the instance with the first generated local account address. It will have DEFAULT_ADMIN role

```
await meemVite.initialize('0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266')
```

### Set the token URI contract

To properly return the `tokenURI`, the contract of the MeemViteURI address must be set

```
await meemVite.setTokenURIContractAddress(<MeemViteURI address>)
```

### Mint a MeemVite

```
await meemVite.mint('<To address>')
```

### Get meem metadata uri

```
await meemVite.tokenURI(<Meem id>)

await meemVite.tokenURI(0)
```

### Grant a role

Available Roles:

* `DEFAULT_ADMIN_ROLE`
* `PAUSER_ROLE`
* `MINTER_ROLE`
* `UPGRADER_ROLE`

```
await meemVite.grantRole((await meemVite.MINTER_ROLE()), '<address>')
```

## Faucets

### Rinkeby

* https://faucet.rinkeby.io/
* https://app.mycrypto.com/faucet
* http://rinkeby-faucet.com/
* https://faucets.blockxlabs.com/ethereum

Offline / Not working consistently
* https://rinkeby.faucet.epirus.io/#


### Ropsten

5 ETH every 24 hours: https://faucet.dimensions.network/

1 ETH but spotty availability: https://faucet.metamask.io/

CHANGELOG.md
## [1.0.1](https://github.com/meemproject/meemvite-nft/compare/v1.0.0...v1.0.1) (2021-10-08)

# 1.0.0 (2021-10-07)


### Features

* implement tests and circleci config ([2fed375](https://github.com/meemproject/meemvite-nft/commit/2fed375))


test/init.test.ts
import { assert } from 'chai'
import { Signer } from 'ethers'
import { ethers } from 'hardhat'
import { MeemVite } from '../typechain/MeemVite'

async function getContract(): Promise<MeemVite> {
	const MV = await ethers.getContractFactory('MeemVite')
	const meemVite = await MV.deploy()
	await meemVite.deployed()

	return meemVite as MeemVite
}

describe('MeemVite', function Test() {
	// let accounts: Signer[]

	// beforeEach(async () => {
	// 	accounts = await ethers.getSigners()
	// })

	it('should do something right', async () => {
		const meemVite = await getContract()
		const [owner, addr1] = await ethers.getSigners()
		console.log({ owner, addr1, ethers })
		// Do something with the accounts
		// console.log({ accounts, name: meemVite.functions })
		// assert.isArray(accounts)
		// assert.isAbove(accounts.length, 1)
	})
})


test/meemVite.test.ts
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import chai, { assert } from 'chai'
import chaiAsPromised from 'chai-as-promised'
import { ethers, waffle } from 'hardhat'
import MeemViteArtifact from '../artifacts/contracts/MeemVite.sol/MeemVite.json'
import MeemViteUriArtifact from '../artifacts/contracts/MeemViteURI.sol/MeemViteURI.json'
import { MeemVite } from '../typechain/MeemVite'
import { MeemViteUri } from '../typechain/MeemViteUri'

chai.use(chaiAsPromised)

const { deployContract } = waffle

describe('MeemVite', function Test() {
	let meemVite: MeemVite
	let meemViteUri: MeemViteUri
	let signers: SignerWithAddress[]

	beforeEach(async () => {
		signers = await ethers.getSigners()
		meemVite = (await deployContract(signers[0], MeemViteArtifact)) as MeemVite
		meemViteUri = (await deployContract(
			signers[0],
			MeemViteUriArtifact
		)) as MeemViteUri
		await meemVite.initialize('0x0000000000000000000000000000000000000000')
		await meemVite.deployed()
		await meemViteUri.deployed()
		await meemVite.setTokenURIContractAddress(meemViteUri.address)
	})

	it('Can mint as owner', async () => {
		const { status } = await (
			await meemVite.connect(signers[0]).mint(signers[0].address)
		).wait()
		assert.equal(status, 1)
	})

	it('Can not mint as non-owner', async () => {
		const isMinter = await meemVite.hasRole(
			await meemVite.MINTER_ROLE(),
			signers[1].address
		)

		assert.isFalse(isMinter)

		await assert.isRejected(
			meemVite.connect(signers[1]).mint(signers[1].address)
		)
	})

	it('Can set URI contract as owner', async () => {
		const { status } = await (
			await meemVite
				.connect(signers[0])
				.setTokenURIContractAddress(meemViteUri.address)
		).wait()

		assert.equal(status, 1)
	})

	it('Can not set URI contract as non-owner', async () => {
		await assert.isRejected(
			meemVite
				.connect(signers[1])
				.setTokenURIContractAddress(meemViteUri.address)
		)
	})

	it('Can get token metadata as anyone', async () => {
		// Mint
		await meemVite.connect(signers[0]).mint(signers[0].address)
		await meemVite.connect(signers[0]).mint(signers[0].address)

		const token0URI = await meemVite.connect(signers[1]).tokenURI(0)
		const token1URI = await meemVite.connect(signers[1]).tokenURI(1)

		const buff0 = Buffer.from(
			token0URI.replace('data:application/json;base64,', ''),
			'base64'
		)

		const json0 = JSON.parse(buff0.toString())
		assert.equal(json0.name, 'MeemVite Seat #0')
		assert.equal(json0.description, 'Join us at https://discord.gg/5NP8PYN8')
		assert.equal(json0.external_url, 'https://meem.wtf/')
		assert.isOk(json0.image)
		assert.equal(json0.background_color, '000000')
		assert.equal(json0.attributes.length, 2)
		const levelAttr0 = json0.attributes.find(
			(a: Record<string, string>) => a.trait_type === 'Level'
		)
		const seatAttr0 = json0.attributes.find(
			(a: Record<string, string>) => a.trait_type === 'Seat'
		)
		assert.equal(levelAttr0.value, 'Ground Floor')
		assert.equal(seatAttr0.value, '0000')

		const buff1 = Buffer.from(
			token1URI.replace('data:application/json;base64,', ''),
			'base64'
		)

		const json1 = JSON.parse(buff1.toString())
		assert.equal(json1.name, 'MeemVite Seat #1')
		assert.equal(json1.description, 'Join us at https://discord.gg/5NP8PYN8')
		assert.equal(json1.external_url, 'https://meem.wtf/')
		assert.isOk(json1.image)
		assert.equal(json1.background_color, '000000')
		assert.equal(json1.attributes.length, 2)
		const levelAttr1 = json1.attributes.find(
			(a: Record<string, string>) => a.trait_type === 'Level'
		)
		const seatAttr1 = json1.attributes.find(
			(a: Record<string, string>) => a.trait_type === 'Seat'
		)
		assert.equal(levelAttr1.value, 'Ground Floor')
		assert.equal(seatAttr1.value, '0001')
	})

	it('Can set role as owner and then mint w/ newly granted account', async () => {
		const { status } = await (
			await meemVite
				.connect(signers[0])
				.grantRole(await meemVite.MINTER_ROLE(), signers[1].address)
		).wait()

		assert.equal(status, 1)

		await meemVite.connect(signers[1]).mint(signers[1].address)
	})

	it('Can not set role as non-owner', async () => {
		await assert.isRejected(
			meemVite
				.connect(signers[1])
				.grantRole(await meemVite.MINTER_ROLE(), signers[2].address)
		)
	})

	it('Can set contractURI as owner', async () => {
		const { status } = await (
			await meemVite
				.connect(signers[0])
				.setContractURI(
					'{"name": "MeemVite Updated","description": "Meems are pieces of digital content wrapped in more advanced dynamic property rights. They are ideas, stories, images -- existing independently from any social platform -- whose creators have set the terms by which others can access, remix, and share in their value. Join us at https://discord.gg/5NP8PYN8","image": "https://meem-assets.s3.amazonaws.com/meem.jpg","external_link": "https://meem.wtf","seller_fee_basis_points": 1000, "fee_recipient": "0x40c6BeE45d94063c5B05144489cd8A9879899592"}'
				)
		).wait()

		assert.equal(status, 1)

		const contractURI = await meemVite.connect(signers[0]).contractURI()
		const buff = Buffer.from(
			contractURI.replace('data:application/json;base64,', ''),
			'base64'
		)

		const json = JSON.parse(buff.toString())
		assert.equal(json.name, 'MeemVite Updated')
	})

	it('Can not set contractURI as non-owner', async () => {
		await assert.isRejected(
			meemVite
				.connect(signers[1])
				.setContractURI(
					'{"name": "MeemVite Updated","description": "Meems are pieces of digital content wrapped in more advanced dynamic property rights. They are ideas, stories, images -- existing independently from any social platform -- whose creators have set the terms by which others can access, remix, and share in their value. Join us at https://discord.gg/5NP8PYN8","image": "https://meem-assets.s3.amazonaws.com/meem.jpg","external_link": "https://meem.wtf","seller_fee_basis_points": 1000, "fee_recipient": "0x40c6BeE45d94063c5B05144489cd8A9879899592"}'
				)
		)
	})

	it('Can get contractURI as non-owner', async () => {
		const contractURI = await meemVite.connect(signers[0]).contractURI()
		const buff = Buffer.from(
			contractURI.replace('data:application/json;base64,', ''),
			'base64'
		)

		const json = JSON.parse(buff.toString())
		assert.equal(json.name, 'MeemVite')
	})

	it('Can set proxy address as owner', async () => {
		const { status } = await (
			await meemVite
				.connect(signers[0])
				.setProxyRegistryAddress('0x58807baD0B376efc12F5AD86aAc70E78ed67deaE')
		).wait()

		assert.equal(status, 1)
	})

	it('Can not set proxy address non-owner', async () => {
		await assert.isRejected(
			meemVite
				.connect(signers[1])
				.setProxyRegistryAddress('0x58807baD0B376efc12F5AD86aAc70E78ed67deaE')
		)
	})
})


hardhat.config.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import dotenv from 'dotenv'
import { HardhatUserConfig } from 'hardhat/config'
import '@nomiclabs/hardhat-waffle'
import '@nomiclabs/hardhat-etherscan'
import '@float-capital/solidity-coverage'
import 'hardhat-typechain'
import 'hardhat-abi-exporter'
import '@openzeppelin/hardhat-upgrades'
import 'hardhat-gas-reporter'
import 'hardhat-contract-sizer'
import './tasks'

dotenv.config()

const config: HardhatUserConfig = {
	solidity: {
		compilers: [
			{ version: '0.8.4' },
			{ version: '0.8.3' },
			{ version: '0.8.2' }
		],
		settings: {
			optimizer: {
				enabled: true,
				runs: 10_000
			}
		}
	},
	defaultNetwork: 'local',
	networks: {
		mainnet: {
			url: `https://mainnet.infura.io/v3/${process.env.INFURA_PROJECT_ID}`,
			accounts: [process.env.LIVE_WALLET_PRIVATE_KEY!].filter(Boolean)
		},
		xdai: {
			timeout: 120000,
			url: `https://xdai.poanetwork.dev`,
			accounts: process.env.TESTNET_MNEMONIC
				? { mnemonic: process.env.TESTNET_MNEMONIC }
				: [process.env.TESTNET_WALLET_PRIVATE_KEY!].filter(Boolean)
		},
		polygon: {
			timeout: 120000,
			url: 'https://polygon-rpc.com',
			// url: 'https://matic-mainnet-full-rpc.bwarelabs.com',
			// url: 'https://polygon-mainnet.g.alchemy.com/v2/xLwwfjFEFLvv_mRhnv7ZW3qM8f3K8MHE',
			// url: 'https://rpc-mainnet.maticvigil.com',
			// url: 'https://matic-mainnet.chainstacklabs.com',
			// chainId: 137,
			// gasPrice,
			// gasPrice: 40000000000,
			// gas: 75725145187,
			// gasMultiplier: 1.5,
			accounts: process.env.LIVE_MNEMONIC
				? { mnemonic: process.env.LIVE_MNEMONIC }
				: [process.env.LIVE_WALLET_PRIVATE_KEY!].filter(Boolean)
		},
		mumbai: {
			url: 'https://rpc-mumbai.maticvigil.com',
			chainId: 80001,
			gasPrice: 10000000000,
			accounts: process.env.TESTNET_MNEMONIC
				? { mnemonic: process.env.TESTNET_MNEMONIC }
				: [process.env.TESTNET_WALLET_PRIVATE_KEY!].filter(Boolean)
		},
		rinkeby: {
			url: `https://rinkeby.infura.io/v3/${process.env.INFURA_PROJECT_ID}`,
			accounts: process.env.TESTNET_MNEMONIC
				? { mnemonic: process.env.TESTNET_MNEMONIC }
				: [process.env.TESTNET_WALLET_PRIVATE_KEY!].filter(Boolean)
		},
		ropsten: {
			url: `https://ropsten.infura.io/v3/${process.env.INFURA_PROJECT_ID}`,
			accounts: process.env.TESTNET_MNEMONIC
				? { mnemonic: process.env.TESTNET_MNEMONIC }
				: [process.env.TESTNET_WALLET_PRIVATE_KEY!].filter(Boolean)
		},
		local: {
			url: 'http://127.0.0.1:8545',
			// gasPrice:
			// 	typeof process.env.LOCAL_GAS_PRICE !== 'undefined'
			// 		? +process.env.LOCAL_GAS_PRICE
			// 		: 65,
			accounts: process.env.LOCAL_MNEMONIC
				? { mnemonic: process.env.LOCAL_MNEMONIC }
				: [process.env.LOCAL_WALLET_PRIVATE_KEY!].filter(Boolean)
		}
	},
	etherscan: {
		apiKey: process.env.ETHERSCAN_API_KEY
	},
	abiExporter: {
		path: './abi',
		clear: true
	},
	gasReporter: {
		enabled: !process.env.CI,
		currency: 'USD',
		gasPrice: 66,
		src: 'contracts',
		coinmarketcap: '7643dfc7-a58f-46af-8314-2db32bdd18ba'
	},
	mocha: {
		timeout: 60_000
	}
}
export default config


tasks/deployURI.ts
import { task } from 'hardhat/config'

type ContractName = 'MeemViteURI'

interface Contract {
	args?: (string | number | (() => string | undefined))[]
	address?: string
	libraries?: () => Record<string, string>
	waitForConfirmation?: boolean
}

task('deployURI', 'Deploys MeemViteURI').setAction(async (args, { ethers }) => {
	const [deployer] = await ethers.getSigners()
	console.log('Deploying contracts with the account:', deployer.address)

	console.log('Account balance:', (await deployer.getBalance()).toString())

	// const nonce = await deployer.getTransactionCount();
	const contracts: Record<ContractName, Contract> = {
		MeemViteURI: {}
	}

	const MeemViteURI = await ethers.getContractFactory('MeemViteURI')

	const meemViteURI = await MeemViteURI.deploy()

	await meemViteURI.deployed()

	console.log('MeemViteURI deployed to:', meemViteURI.address)

	return contracts
})


tasks/upgradeContract.ts
import { task, types } from 'hardhat/config'

type ContractName = 'Meem'

interface Contract {
	args?: (string | number | (() => string | undefined))[]
	address?: string
	libraries?: () => Record<string, string>
	waitForConfirmation?: boolean
}

task('upgradeContract', 'Deploys Meem Gate Token')
	.addParam(
		'contractaddress',
		'The originally deployed contract address',
		undefined,
		types.string,
		false
	)
	.setAction(async (args, { ethers, upgrades }) => {
		const [deployer] = await ethers.getSigners()
		console.log('Deploying contracts with the account:', deployer.address)

		console.log('Account balance:', (await deployer.getBalance()).toString())

		const contracts: Record<ContractName, Contract> = {
			Meem: {}
		}

		const MeemVite = await ethers.getContractFactory('MeemVite')
		const meemVite = await upgrades.upgradeProxy(args.contractaddress, MeemVite)

		await meemVite.deployed()

		console.log('MeemVite deployed to:', meemVite.address)

		return contracts
	})


tasks/deploy.ts
import { task } from 'hardhat/config'

type ContractName = 'MeemVite'

interface Contract {
	args?: (string | number | (() => string | undefined))[]
	address?: string
	libraries?: () => Record<string, string>
	waitForConfirmation?: boolean
}

task('deploy', 'Deploys MeemVite').setAction(
	async (args, { ethers, upgrades, hardhatArguments }) => {
		const [deployer] = await ethers.getSigners()
		console.log('Deploying contracts with the account:', deployer.address)

		console.log('Account balance:', (await deployer.getBalance()).toString())

		// const nonce = await deployer.getTransactionCount();
		const contracts: Record<ContractName, Contract> = {
			MeemVite: {}
		}

		// This is the OpenSea proxy address which will allow trading to work properly
		let proxyRegistryAddress = ''

		switch (hardhatArguments.network) {
			case 'matic':
			case 'polygon':
				proxyRegistryAddress = '0x58807baD0B376efc12F5AD86aAc70E78ed67deaE'
				break

			case 'rinkeby':
				proxyRegistryAddress = '0xf57b2c51ded3a29e6891aba85459d600256cf317'
				break

			case 'mainnet':
				proxyRegistryAddress = '0xa5409ec958c83c3f309868babaca7c86dcb077c1'
				break

			case 'local':
				proxyRegistryAddress = '0x0000000000000000000000000000000000000000'
				break

			default:
				break
		}

		const MeemVite = await ethers.getContractFactory('MeemVite')

		const meemVite = await upgrades.deployProxy(
			MeemVite,
			[proxyRegistryAddress],
			{
				kind: 'uups'
			}
		)

		await meemVite.deployed()

		console.log('MeemVite deployed to:', meemVite.address)

		return contracts
	}
)
